File: packages/penca-client/src/penca_client/status.py
"""Client-side gRPC status translation.

Mirror of ``penca_server_grpc.status`` — where the server maps
``ApiError`` subclasses onto ``grpc.StatusCode`` values, this module does
the inverse: takes a ``grpc.RpcError`` raised by a stub call and returns
the matching ``ApiError`` subclass so callers see a typed Python
exception rather than a bare gRPC error.
"""

from __future__ import annotations

from grpc import RpcError, StatusCode

from penca_client.errors import (
    AbortedError,
    AlreadyExistsError,
    ApiError,
    FailedPreconditionError,
    InvalidRequestError,
    NotFoundError,
    QueryTimeoutError,
)


def rpc_error_to_api_error(err: RpcError) -> Exception:
    """Convert a gRPC ``RpcError`` into an idiomatic Python exception.

    Most statuses map to an ``ApiError`` subclass; ``UNIMPLEMENTED``
    maps to Python's built-in ``NotImplementedError`` so callers can
    tell "not wired up yet" from "internal failure".

    The concrete call objects grpcio raises inherit from both
    ``RpcError`` and ``Call``; ``Call`` is what provides ``code()`` /
    ``details()``, but the type stubs only expose the bare ``RpcError``
    base.
    """
    code = err.code()  # ty: ignore[unresolved-attribute]
    detail = err.details() or str(err)  # ty: ignore[unresolved-attribute]
    if code == StatusCode.INVALID_ARGUMENT:
        return InvalidRequestError(detail)

    if code == StatusCode.NOT_FOUND:
        return NotFoundError(detail)

    if code == StatusCode.FAILED_PRECONDITION:
        return FailedPreconditionError(detail)

    if code == StatusCode.ALREADY_EXISTS:
        return AlreadyExistsError(detail)

    if code == StatusCode.UNIMPLEMENTED:
        return NotImplementedError(detail)

    # ADR 0019 / CHA-233: `read_data` / `audit_data` past
    # ``query_timeout_seconds`` surfaces as ``RESOURCE_EXHAUSTED``;
    # mirror the Rust ``ApiError::QueryTimeout`` variant so callers
    # can catch a typed Python exception.
    # ABORTED is the concurrency-conflict code: the operation made no changes,
    # so reissuing is safe. Mapped to its own class so a caller can retry it
    # without also swallowing genuine failures.
    if code == StatusCode.ABORTED:
        return AbortedError(detail)

    if code == StatusCode.RESOURCE_EXHAUSTED:
        return QueryTimeoutError(detail)

    return ApiError(detail)

File: crates/penca-server-grpc/src/status.rs
//! Error mapping from `ApiError` to `tonic::Status`.

use penca_api::ApiError;
use tonic::Status;

/// Convert an [`ApiError`] into the appropriate [`tonic::Status`] code.
pub fn api_error_to_status(err: ApiError) -> Status {
    match &err {
        ApiError::InvalidRequest(_) => Status::invalid_argument(err.to_string()),
        ApiError::NotFound(_) => Status::not_found(err.to_string()),
        // CHA-236 — name-uniqueness collision on Create* / Update*.
        ApiError::AlreadyExists(_) => Status::already_exists(err.to_string()),
        ApiError::FailedPrecondition(_) => Status::failed_precondition(err.to_string()),
        // A precondition raised down in the metadata layer is still a
        // precondition. Without this it falls through to the `_` arm and reaches
        // the client as INTERNAL, which reads like a server bug for something the
        // caller can fix by changing the request.
        ApiError::Metadata(penca_storage_meta::MetadataError::FailedPrecondition(_)) => {
            Status::failed_precondition(err.to_string())
        }
        ApiError::Unimplemented(_) => Status::unimplemented(err.to_string()),
        // ADR 0019 §"Defaults" — the cap surfaces as RESOURCE_EXHAUSTED;
        // the message body already names `query_timeout_seconds` and
        // the retry pattern.
        ApiError::QueryTimeout(_) => Status::resource_exhausted(err.to_string()),
        // `Aborted` is the canonical "concurrency conflict, retry at a higher
        // level" code — the distinction a caller needs to tell a lost lock race
        // from a genuine failure.
        ApiError::Aborted(_) => Status::aborted(err.to_string()),
        _ => Status::internal(err.to_string()),
    }
}
